packages/metadata-editor/src/MetadataEditor.tsx

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright 2018-2020 IBM Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { FormGroup } from '@blueprintjs/core';
+
+import { FrontendServices } from '@elyra/application';
+import { DropDown } from '@elyra/ui-components';
+
+import {
+  ReactWidget,
+  WidgetTracker,
+  showDialog,
+  Dialog
+} from '@jupyterlab/apputils';
+import { CodeEditor } from '@jupyterlab/codeeditor';
+import { InputGroup, Button } from '@jupyterlab/ui-components';
+
+import { Message } from '@lumino/messaging';
+
+import * as React from 'react';
+
+import { METADATA_EDITOR_ID } from './index';
+
+const ELYRA_METADATA_EDITOR_CLASS = 'elyra-metadataEditor';
+const DIRTY_CLASS = 'jp-mod-dirty';
+
+type FormItemType = 'TextInput' | 'DropDown' | 'Code';
+
+export type FormItem = {
+  value: any;
+  type: FormItemType;
+  label: string;
+  schemaField: string;
+};
+
+/**
+ * Metadata editor widget
+ */
+export class MetadataEditor extends ReactWidget {
+  metadata: FormItem[];
+  newFile: boolean;
+  updateSignal: any;
+  fileName: string;
+  editorFactory: CodeEditor.Factory;
+  editor: CodeEditor.IEditor;
+  namespace: string;
+  tracker: WidgetTracker;
+  dirty: boolean;
+
+  constructor(
+    metadata: FormItem[],
+    newFile: boolean,
+    updateSignal: any,
+    editorFactory: CodeEditor.Factory | null,
+    namespace: string,
+    tracker: WidgetTracker,
+    fileName?: string
+  ) {
+    super();
+    this.metadata = metadata;
+    this.editorFactory = editorFactory;
+    this.namespace = namespace;
+    this.newFile = newFile;
+    this.updateSignal = updateSignal;
+    this.handleTextInputChange = this.handleTextInputChange.bind(this);
+    this.handleDropdownChange = this.handleDropdownChange.bind(this);
+    this.tracker = tracker;
+    if (fileName) {
+      this.fileName = fileName;
```

Comment:
What is fileName?  Should this be called `name`?  Even though the server equates `name` to a file, we don't ever send the filename over the REST API, so I'm not sure what `fileName` is referring to.

Reply:
Yeah, I'm just gonna remove this field, you're right.